### PR TITLE
Add "Replace question" action to dashboard cards

### LINKED ISSUE #36497: Let creators swap out dashboard questions with a different question
## Context

- [Product doc](https://www.notion.so/metabase/Let-creators-swap-out-dashboard-cards-with-something-else-1cbb54963f06471885d5bc2cea608dcb)
- #36586

The project seems small enough not to need a feature branch

## Tasks

- [x] #36589
- [ ] Add "Replace card" feature

### LINKED ISSUE #36586: Testing plan for "Let creators swap out dashboard questions with a different question"
For epic #36497

## Dimensions

* dashcard
	* type: query, text, link, etc.
	* tab: 1st, 2nd, 3rd, etc.
* old question
	* has filters mapped
	* has visualization settings overrides
	* has click behavior
	* has added series
* new question
	* permissions: full, view only, no data
## Actions

* replace card with another card
* edit the dashcard after replacing a question
* cancel (by closing the question picker)
* undo replace
### Unit and integration tests

 1. `DashCard`
	 1. should show the "replace" action for dashcards with questions
	 2. should show the "replace" action for dashcards with models
	 3. should not show the "replace" action for heading, text, and link dashcards
 2. `QuestionPickerModal`
	 1. should list questions
	 2. should navigate through collections
	 3. should search for questions
	 4. should close
	 5. should call `onSelect` and close
 3. `replaceCard` action
	 1. should change dashcard's `card_id` and `card` properties
	 2. should run new card query (should use the `/card/:id/query` endpoint)
	 4. should reset card's previous
		 1. `series`
		 2. `parameter_mappings`
		 3. `visualization_settings`
		 4. click behavior
	 5. should auto-wire parameters to new card
	 6. should not change tab ID
### E2E tests

It seems like most things (except for "Undo") can be covered on unit / integration test levels.

So in Cypress we can test the happy path and a few more complex scenarios.

**Happy path**

1. Enter dashboard edit mode
2. Replace dashcard's card with another card
3. Ensure the query reruns and the expected visualization and results are shown
4. Ensure parameters auto-wiring worked as expected
5. Ensure visualization settings, click behavior, etc. are clean after the swap
6. Ensure changes are persisted after clicking "Save"

**More complex scenarios**

* ensure "Undo replace" doesn't just revert the dashcard to the "pre-edit" state
	* edit dashboard > change visualization settings > replace card > undo > ensure visualization setting changes are still there
* ensure we're handling cards with view-only and "no data" permissions

### PR DESCRIPTION
**Pull request by kulyk**
Closes #36497, closes #36586

Adds a "Replace" button to dashboard cards with questions or models. It makes it possible to swap the dashboard card's query while keeping the dashboard layout intact. It's important to call out that we would reset properties of a dash card that no longer makes sense with a new query: parameter mappings, click behavior, visualization settings, and added series.

### How to verify

1. Open any dashboard that has a few cards
2. Enter the editing mode
3. Hover a dashboard card using a question or a model
4. Find and click the "Replace" action in the top-right floating panel
5. Select a new question or a model from the modal picker
6. Ensure the dashboard card's content is replaced, and the query is rerun
7. Ensure parameters auto-wiring worked (your dashboard should have another card using the same table and mapped to some parameters)
8. Ensure you can undo the revert
9. Ensure changes are persisted after clicking "Save"

### Demo

<details>
<summary>Basic</summary>

https://github.com/metabase/metabase/assets/17258145/15363bc3-7a37-49f4-90b6-b3af2096c842

</details>

<details>
<summary>Filters auto-wiring</summary>

https://github.com/metabase/metabase/assets/17258145/38d75e0f-2185-470a-926f-bca7ab74c498

</details>

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Review by cdeweyx — APPROVED** [review]
Working like a charm 👍

**Review by kdoh — APPROVED** [review]

**Review by snoe — APPROVED** [review]
BE LGTM

**Comment by calherries** [comment]
If Case is cool with updating card IDs, you should change the logic in `models.dashboard-card.update-dashboard-card!` which prevents it https://github.com/metabase/metabase/blob/ffe0385b01b698b02d66b9d18b4dcc238b200d7d/src/metabase/models/dashboard_card.clj#L182-L184

**Comment by markbastian** [comment]
> If Case is cool with updating card IDs, you should change the logic in `models.dashboard-card.update-dashboard-card!` which prevents it
> 
> https://github.com/metabase/metabase/blob/ffe0385b01b698b02d66b9d18b4dcc238b200d7d/src/metabase/models/dashboard_card.clj#L182-L184

Isn't this addressed here? https://github.com/metabase/metabase/pull/36744/files#diff-b6abb82aa02f4647f6ac74d17bfa75dda523bf75c2eed11954c56127a6d7446dR180-R189

**Comment by alxnddr** [comment]
It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
Screenshot 2023-12-13 at 6 43 34 PM

File: frontend/src/metabase/dashboard/selectors/selectors-typed.ts [hunk 1]
```diff
@@ -5,6 +5,10 @@ export function getDashboardId(state: State) {
   return state.dashboard.dashboardId;
 }
 
+export function getDashboardBeforeEditing(state: State) {
+  return state.dashboard.isEditing;
```
Comment (alxnddr) [inline frontend/src/metabase/dashboard/selectors/selectors-typed.ts:9]: Oh, `isEditing` is a dashboard instead of a boolean 🤯

**Review by alxnddr — APPROVED** [review]
Nice and clean! Please have a look at the suggestion about errored cards

File: frontend/src/metabase/dashboard/selectors/selectors-typed.ts (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/dashboard/selectors/selectors-typed.ts:9]: I'm afraid you're right 😢

**Comment by kulyk** [comment]
> It would be great if users could replace errored cards, I think it can be a valid use case for the functionality

@alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!

https://github.com/metabase/metabase/assets/17258145/1b1085fd-99ba-4752-bcd6-5b4df064aa24

**Comment by perivamsi** [comment]
> > It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
> 
> @alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!
> 
> Demo

If it is possible to show the "replace card" prompt directly on the error card, that will be helpful.

So change "There was a problem displaying this card." by adding a "replace it" link or button at the end? This is more discoverable.

**Comment by kulyk** [comment]
> > > It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
> 
> > 
> 
> > @alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!
> 
> > 
> 
> > Demo
> 
> 
> 
> If it is possible to show the "replace card" prompt directly on the error card, that will be helpful.
> 
> 
> 
> So change "There was a problem displaying this card." by adding a "replace it" link or button at the end? This is more discoverable.

For context, we figured we'd prefer to keep the action as is. People might prefer to fix the issue instead of replacing the card

**Pull request merged**